test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/DeepInsertE2ETests.cs

```diff
@@ -0,0 +1,715 @@
+                {""ID"":300,""Name"":""Person 300"", ""Cars"":[{""ID"":1003,""Name"":""Car 1003"",""Manufacturers"":[{""ID"":101,""Name"":""ManufacturerC""}], ""Owners"":[{""ID"":1,""Name"":""OwnerA""}]}]}
+            ]}";
+
+        private const string cars = @"{
+            ""@odata.context"":""http://localhost:5000/$metadata#Cars"",
+            ""value"":[
+                {""ID"":1001,""Name"":""Car 1001"",""Manufacturers"":[{""ID"":101,""Name"":""ManufacturerA""}], ""Owners"":[{""ID"":1,""Name"":""OwnerA""}]},
+                {""ID"":1002,""Name"":""Car 1002"",""Manufacturers"":[{""ID"":101,""Name"":""ManufacturerB""}], ""Owners"":[{""ID"":1,""Name"":""OwnerA""}]},
+                {""ID"":1003,""Name"":""Car 1003"",""Manufacturers"":[{""ID"":101,""Name"":""ManufacturerC""}], ""Owners"":[{""ID"":1,""Name"":""OwnerA""}]}
+            ]}";
+
+        private readonly Container context;
+        private readonly RequestInfo requestInfo;
+        private readonly Serializer serializer;
+        private readonly HeaderCollection headers;
+
+        public DeepInsertE2ETests()
+        {
+            this.context = new Container(new Uri("http://localhost:5000"));
+            this.requestInfo = new RequestInfo(context);
+            this.serializer = new Serializer(this.requestInfo);
+            this.headers = new HeaderCollection();
+        }
+
+        [Fact]
+        public void CallingDeepInsert_WithNullArguments_ShouldThrowAnException()
+        {
+            Assert.Throws<ArgumentNullException>(() => this.context.DeepInsert<Person>(null));
+        }
+
+        [Fact]
+        public void CallingDeepInsertRequest_WithNullArguments_ShouldThrowAnException()
+        {
+            var result = new DeepInsertSaveResult(this.context, Util.SaveChangesMethodName, SaveChangesOptions.DeepInsert, null, null);
+            Assert.Throws<ArgumentNullException>(() => result.DeepInsertRequest<Person>(null));
+        }
+
+        [Fact]
+        public void DeepInsertAnEntry_WithOneLevelOfNesting()
+        {
+            var expectedResponse = "{\"@context\":\"http://localhost:5000/$metadata#Persons(Cars())/$entity\",\"ID\":100,\"Name\":\"Person 100\",\"Cars\":[{\"ID\":\"1001\",\"Name\":\"Car 1001\"}]}";
+            var headers = new Dictionary<string, string>()
+                        {
+                            {"Content-Type", "application/json;charset=utf-8"},
+                            {"Location", "http://localhost:5000/Persons(100)"},
+                        };
+
+            SetupContextWithRequestPipelineForSaving(
+                this.context,
+                expectedResponse,
+                headers);
+
+            var person = new Person
+            {
+                ID = 100,
+                Name = "Person 100",
+            };
+
+            var car = new Car
+            {
+                ID = 1001,
+                Name = "Car 1001"
+            };
+
+            this.context.AddObject("Persons", person);

```

Comment:
What happens when you insert an object without an ID and the ID is generated by the server? Will the response contain the updated value? Will this be reflected in the entity tracker?

Reply:
The response has the returned ID.
I have added the test.

In the implementation, we update the entry entitydescriptor with the materialized object. We then apply the updated descriptor to the entity tracker.
https://github.com/OData/odata.net/blob/34499522b6a3087aad1bb7568ed2ab97f7961a86/src/Microsoft.OData.Client/Materialization/MaterializerEntry.cs#L305
https://github.com/OData/odata.net/blob/34499522b6a3087aad1bb7568ed2ab97f7961a86/src/Microsoft.OData.Client/AtomMaterializerLog.cs#L180